### PR TITLE
CDAP-13758 fixing latest status persistence by syncing file length ev…

### PR DESCRIPTION
**Pull request by gsps1**
…ery 10 seconds

By using update length flag, when hsync is called, the length of the file is also updated.
This fixes issue we were noticing with reading the latest avro file where all the contents were not read by the spark application.
in order to not overload, we only perform hsync every 10 seconds.

### CODE REVIEW

File: cdap-app-templates/cdap-program-report/cdap-program-report-base/src/main/java/co/cask/cdap/report/main/RunMetaFileManager.java [hunk 1]
```diff
@@ -41,24 +42,30 @@
   private static final Long DEFAULT_MAX_FILE_OPEN_DURATION = TimeUnit.HOURS.toMillis(6);
   private static final String SYNC_INTERVAL = "file.sync.interval.bytes";
   private static final String MAX_FILE_SIZE_BYTES = "file.max.size.bytes";
-  private static final String MAX_FILE_OPEN_DURATION = "file.max.open.duration";
+  private static final String MAX_FILE_OPEN_DURATION_MILLIS = "file.max.open.duration.millis";
```
Comment (sreevatsanraman) [inline cdap-app-templates/cdap-program-report/cdap-program-report-base/src/main/java/co/cask/cdap/report/main/RunMetaFileManager.java:46]: It is not file open duration right? It is the sync interval?

Comment (gsps1) [reply cdap-app-templates/cdap-program-report/cdap-program-report-base/src/main/java/co/cask/cdap/report/main/RunMetaFileManager.java:46]: that is a separate property. `SYNC_INTERVAL_TIME_MILLIS` this change is an improvement in naming of an existing property to include time unit.

File: cdap-app-templates/cdap-program-report/cdap-program-report-base/src/main/java/co/cask/cdap/report/main/RunMetaFileManager.java [hunk 2]
```diff
@@ -76,16 +83,25 @@ public void append(ProgramRunInfo programRunInfo) throws InterruptedException {
                                programRunInfo.getNamespace(), programRunInfo.getTimestamp());
     RunMetaFileOutputStream outputStream = namespaceToLogFileStreamMap.get(programRunInfo.getNamespace());
     appendAndFlushWithRetry(outputStream, programRunInfo);
+    syncOutputStreamsIfRequired();
```
Comment (sreevatsanraman) [inline cdap-app-templates/cdap-program-report/cdap-program-report-base/src/main/java/co/cask/cdap/report/main/RunMetaFileManager.java:86]: We only sync if there are new entries. So say if there are five events

t, t+1sec, t+8secs,t+20secs,t+1hour 

the fourth event will not be seen until the 1 hour event arrives. Instead can we make it sync on a periodic basis?

File: cdap-app-templates/cdap-program-report/cdap-program-report-base/src/main/java/co/cask/cdap/report/main/RunMetaFileManager.java [hunk 3]
```diff
@@ -41,24 +42,30 @@
   private static final Long DEFAULT_MAX_FILE_OPEN_DURATION = TimeUnit.HOURS.toMillis(6);
   private static final String SYNC_INTERVAL = "file.sync.interval.bytes";
   private static final String MAX_FILE_SIZE_BYTES = "file.max.size.bytes";
-  private static final String MAX_FILE_OPEN_DURATION = "file.max.open.duration";
+  private static final String MAX_FILE_OPEN_DURATION_MILLIS = "file.max.open.duration.millis";
+  private static final long SYNC_INTERVAL_TIME_MILLIS = TimeUnit.SECONDS.toMillis(10);
+  private static final String SYNC_INTERVAL_TIME_MILLIS_METRIC = "last.sync.time.millis";
```
Comment (sreevatsanraman) [inline cdap-app-templates/cdap-program-report/cdap-program-report-base/src/main/java/co/cask/cdap/report/main/RunMetaFileManager.java]: This is specific to run meta right? Can we name it appropriately? suggest: run.metadata.last.sync.time.mills?

File: cdap-app-templates/cdap-program-report/cdap-program-report-base/src/main/java/co/cask/cdap/report/main/RunMetaFileManager.java (hunk 2 above)
Comment (gsps1) [reply cdap-app-templates/cdap-program-report/cdap-program-report-base/src/main/java/co/cask/cdap/report/main/RunMetaFileManager.java:86]: we do call this on periodic basis. as this is also called from https://github.com/caskdata/cdap/blob/fix/CDAP-13758-fixing-latest-status/cdap-app-templates/cdap-program-report/cdap-program-report-base/src/main/java/co/cask/cdap/report/main/TMSSubscriber.java#L82 which is called when there are no events read from TMS.

Comment (gsps1) [reply cdap-app-templates/cdap-program-report/cdap-program-report-base/src/main/java/co/cask/cdap/report/main/RunMetaFileManager.java:86]: i will add a comment in the code.

**Comment by gsps1** [comment]
build green : https://builds.cask.co/browse/CDAP-RUT1547-2

**Review by sreevatsanraman — APPROVED** [review]

**Pull request merged**